### PR TITLE
Fix(RuleRight): Take db groups into account for rights rules on LDAP users

### PR DESCRIPTION
**Pull request by stonebuzz**
Maybe an alternative to 
https://github.com/glpi-project/glpi/pull/16154

The "groups" criteria for authorizations rules is not processed correctly for LDAP users.

| Q             | A
| ------------- | ---
| Bug fix?      | yes/no
| New feature?  | yes/no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !31311

### CODE REVIEW

File: tests/LDAP/AuthLdap.php.orig
Comment (cedric-anne) [inline tests/LDAP/AuthLdap.php.orig:1]: This file should be removed.

File: tests/LDAP/AuthLdap.php [hunk 1]
```diff
@@ -2181,6 +2184,57 @@ public function testRuleRight()
...
+        // Add group to user
+        $this->createItem(Group_User::class, [
+            'groups_id' => $group_id,
+            'users_id'  => $users_id,
+        ]);
+        $groups = (new Group_User())->find([
+            'groups_id' => $group_id,
+            'users_id'  => $users_id,
+        ]);
+        $this->array($groups)->hasSize(1);
```
Comment (cedric-anne) [inline tests/LDAP/AuthLdap.php:2224]: This is useless, `$this->createItem()` already checkes that teh creation is OK.
```suggestion
```

File: tests/LDAP/AuthLdap.php [hunk 2]
```diff
@@ -2181,6 +2184,57 @@ public function testRuleRight()
...
+        $builder->addCriteria('_groups_id', \Rule::PATTERN_IS, $group_id);
+        $builder->addAction('assign', 'profiles_id', 4); // Super admin
+        $builder->addAction('assign', 'entities_id', 1);
+        $builder->setEntity(0);
+        $this->createRule($builder);
+
+        // login the user to force a real synchronisation (and creation into DB)
+        $this->login('brazil7', 'password', false);
+        $users_id = \User::getIdByName('brazil7');
+
```
Comment (cedric-anne) [inline tests/LDAP/AuthLdap.php:2214]:
```suggestion
        // Check that the use is not attached to the profile at creation
        $rights = (new Profile_User())->find([
            'users_id' => $users_id,
            'profiles_id' => 4,
        ]);

        $this->array($rights)->hasSize(0);

```

**Review by cedric-anne — APPROVED** [review]
Seems OK, but I wonder if side effects are sufficiently limited to introduce this change in a bugfix version.

File: tests/LDAP/AuthLdap.php [hunk 3]
```diff
@@ -2181,6 +2184,59 @@ public function testRuleRight()
         $this->array($gus)->hasSize(1);
     }
 
+        /**
```
Comment (AdrienClairembault) [inline tests/LDAP/AuthLdap.php:2187]: Indent looks wrong

File: src/User.php [hunk 4]
```diff
@@ -571,6 +571,26 @@ public function getFromDBbyDn($user_dn)
         ]);
     }
 
+        /**
```
Comment (AdrienClairembault) [inline src/User.php:574]: Indent looks wrong.

File: src/User.php [hunk 5]
```diff
@@ -571,6 +571,26 @@ public function getFromDBbyDn($user_dn)
...
+     * Retrieve a user from the database using it's dn.
+     *
+     * @param string $user_dn dn of the user
```
Comment (AdrienClairembault) [inline src/User.php:577]: The second parameter is missing from the PHPDoc

**Review by AdrienClairembault — CHANGES_REQUESTED** [review]
Looks good to me besides some minor requested changes 👍.

No opinion on the main / bugfixes debate (I'd argue that this is a bug even if technically it was just never implemented).

File: src/User.php [hunk 6]
```diff
@@ -571,6 +571,26 @@ public function getFromDBbyDn($user_dn)
...
+     *
+     * @return boolean
+     */
```
Comment (cedric-anne) [inline src/User.php:580]:
```suggestion
    /**
     * Retrieve a user from the database using it's dn.
     *
     * @param string $user_dn dn of the user
     *
     * @return bool
     */
```

File: tests/LDAP/AuthLdap.php [hunk 7]
```diff
@@ -2181,6 +2184,59 @@ public function testRuleRight()
...
+     * Test if rules targeting ldap criteria are working
+     *
+     * @return void
+     */
```
Comment (cedric-anne) [inline tests/LDAP/AuthLdap.php:2191]:
```suggestion
    /**
     * Test if rules targeting ldap criteria are working
     *
     * @return void
     */
```

File: src/User.php [hunk 8]
```diff
@@ -571,6 +571,26 @@ public function getFromDBbyDn($user_dn)
...
     }
 
+    /**
+     * Retrieve a user from the database using it's dn.
+     *
+     * @param string $user_dn dn of the user
+     *
+     * @return bool
+     */
+    public function getFromDBbyDnAndAuth($user_dn, $auths_id): bool
```
Comment (cedric-anne) [inline src/User.php:581]:
```suggestion
    public function getFromDBbyDnAndAuth(string $user_dn, int $auths_id): bool
```

File: src/User.php [hunk 9]
```diff
@@ -571,6 +571,26 @@ public function getFromDBbyDn($user_dn)
         ]);
     }
 
+    /**
+     * Retrieve a user from the database using it's dn.
+     *
+     * @param string $user_dn dn of the user
```
Comment (cedric-anne) [inline src/User.php:577]:
```suggestion
     * @param string $user_dn
     * @param int $auths_id
```

**Comment by trasher** [comment]
I have absolutely no idea if that can cause side effects... But that's something new; therefore it should target main (I guess we can provide a patch to the original reporter).

**Review by AdrienClairembault — APPROVED** [review]

**Comment by stonebuzz** [comment]
closed in favour of  https://github.com/glpi-project/glpi/pull/16555 (```main```) version

**Pull request closed without merging**